org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/RecordPatternTest.java

```diff
@@ -4686,4 +4686,54 @@ static void foo(Object o) {
 				"Syntax error on tokens, delete these tokens\n" +
 				"----------\n");
 	}
+
+	public void testIssue3066() {
+		runNegativeTest(new String[] {
+				"X.java",
+				"""
+				public record X<T>(int x) {
+					public static void main(String[] args) {
+						Object o = new Object();
+						switch (o) {
+						case X<String>(int x):
+						default:
+						}
+					}
+				}
+				"""
+			},
+			"""
+			----------
+			1. ERROR in X.java (at line 5)
+				case X<String>(int x):
+				     ^^^^^^^^^^^^^^^^
+			Type X<String> cannot be safely cast to Object
+			----------
```

Comment:
This looks backwards - no ? The error is that Object cannot be safely cast to X<String> and not the other way around ??

Reply:
ups, right, we keep confusing ourselves by naming parameters as "right" and "left", which was ok-ish, while only one syntax was involved, but in pattern matching, where is right, where is left??
(ironically, I just fixed the same kind of bug a few lines up, and then introduced a new instanced of the bug here).
Next commit will include a bit of renaming.

Reply:
Fixed in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3069/commits/bc53322ad7d9e8d9e2ceabcac50db530e08b6e45